nixos/modules/services/networking/teleport.nix

```diff
@@ -0,0 +1,76 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.teleport;
+
+  # Pretty-print JSON to a file
+  writePrettyJSON = name: x:
+    pkgs.runCommand name { } ''
+      echo '${builtins.toJSON x}' | ${pkgs.jq}/bin/jq . > $out
+    '';
+  # This becomes the main config file
+  # To be used in the future when I add seperate option sections for each part of teleport
+  # Rather then writing a single yaml file directly into your nix file.
+  telConfig = {
+    global = cfg.globalConfig;
+    rule_files = cfg.ruleFiles ++ [
+      (pkgs.writeText "teleport.rules" (concatStringsSep "\n" cfg.rules))
+    ];
+  };
+
+  generatedTeleportYml = writePrettyJSON "teleport.yaml" telConfig;
+
+  teleportYml =
+    if cfg.configText != null then
+      pkgs.writeText "teleport.yml" cfg.configText
+    else generatedTeleportYml;
+
+  cmdlineArgs = "-c ${teleportYml}";
+
+in
+
+{
+  options = {
+    services.teleport = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enable the Teleport Cluster SSH daemon.
+        '';
+      };
+      dataDir = mkOption {
+        type = types.path;
+        default = "/var/lib/teleport/";
+        description = ''
+          Directory to store Teleport Service data.
+        '';
+      };
+      configText = mkOption {
+        type = types.nullOr types.lines;
+        default = null;
+        description = ''
+          If non-null, this option defines the text that is written to
+          teleport.yml.
+        '';
+      };
+    };
+  };
+
+    config = mkIf cfg.enable {
+      systemd.services.teleport = {
```

Comment:
There are a few differences here from the services definition from https://gravitational.com/teleport/docs/admin-guide/ that might be worth incorporating.

Reply:
What differences are you thinking of?

Reply:
```
[Unit]
Description=Teleport SSH Service
After=network.target

[Service]
Type=simple
Restart=on-failure
ExecStart=/usr/local/bin/teleport start --config=/etc/teleport.yaml --pid-file=/run/teleport.pid
ExecReload=/bin/kill -HUP $MAINPID
PIDFile=/run/teleport.pid

[Install]
WantedBy=multi-user.target
```
- the lack of an `ExecReload`
- not specifying `--pid-file` in the `ExecStart`

Additionally you should specify `/run` instead of `/var/run`. See https://unix.stackexchange.com/questions/175345/difference-between-run-and-var-run for details.